### PR TITLE
feat: support route hint

### LINKED ISSUE #241: Feature: Support SQL hint to solve no shard query condition cause table full scan.
**What would you like to be added**:

**Why is this needed**:

For this scenes:

`uid` is sharding key for table `student_xxxx`.

If user query data with `uid` condition, `arana` will auto route to the physics table.

But if user query data with no `uid` condition, It will cause table full scan.

Now, we want to support SQL hint, to make SQL route to the table which point out at the hint.

For example:

```
/*+ ARANA: SCAN=employee_*.student_0000 */
```

### PR DESCRIPTION
**Pull request by ZLBer**
**What this PR does**:
support  route hint
**Which issue(s) this PR fixes**:

Fixes #241

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by dongzl — APPROVED** [review]

File: test/integration_test.go [hunk 1]
```diff
@@ -606,3 +606,33 @@ func (s *IntegrationSuite) TestHints() {
...
+		expectLen int
+	}
+
+	for _, it := range []tt{
+		//use route hint
+		{"/*A! route(employees_0000=student_0000,employees_0000=student_0007) */ SELECT * FROM student", 6},
+		//not use route hint, one shard
+		{"/*A! route(employees_0000=student_0000,employees_0000=student_0007) */ SELECT * FROM student where uid=1", 1},
+		//not use route hint, fullScan
+		{"/*A! route(employees_0000=student_0000,employees_0000=student_0007) */ SELECT * FROM student  where uid>=1", 98},
```
Comment (jjeffcaii) [inline test/integration_test.go:626]: How about using `route(employees_0000.student_0000, employees_0000.student_0007)` instead?

Comment (ZLBer) [reply test/integration_test.go]: resolved

File: pkg/runtime/optimize/dml/select.go [hunk 2]
```diff
@@ -87,6 +87,24 @@ func optimizeSelect(ctx context.Context, o *optimize.Optimizer) (proto.Plan, err
 
 	log.Debugf("compute shards: result=%s, isFullScan=%v", shards, fullScan)
 
+	//use route hint when fullScan
+	hints := rcontext.RouteHints(ctx)
```
Comment (sanxun0325) [inline pkg/runtime/optimize/dml/select.go:91]: It can be based https://github.com/arana-db/arana/pull/315 on implement route

Comment (ZLBer) [reply pkg/runtime/optimize/dml/select.go:91]: fine, i will fix after #315 merged

**Comment by ZLBer** [comment]
@jjeffcaii  In this case, it is allowed to use custom route ? 
`/*A! route(employees_0000.student_0000,employees_0000.student_0007) */ SELECT * FROM student where uid=32`

File: pkg/runtime/optimize/hints.go [hunk 3]
```diff
@@ -100,3 +103,21 @@ func (h *Direct) exec(tableName ast.TableName, rule *rule.Rule) (hintTables rule
 	hintTables[db0] = []string{tableName.String()}
 	return hintTables, nil
 }
+
+type CustomRoute struct {
+}
+
+func (c *CustomRoute) exec(tableName ast.TableName, r *rule.Rule, hints []*hint.Hint) (hintTables rule.DatabaseTables, err error) {
+	hintTables = make(map[string][]string)
+	for _, h := range hints {
```
Comment (binbin0325) [inline pkg/runtime/optimize/hints.go:112]: hints should be a collection of hints that contain only the router

File: pkg/runtime/optimize/hints.go [hunk 4]
```diff
@@ -100,3 +103,21 @@ func (h *Direct) exec(tableName ast.TableName, rule *rule.Rule) (hintTables rule
...
+		for _, i := range h.Inputs {
+			tb := strings.Split(i.V, ".")
```
Comment (binbin0325) [inline pkg/runtime/optimize/hints.go:114]: It might be better to validate parameters in validate func

**Comment by jjeffcaii** [comment]
> @jjeffcaii  In this case, it is allowed to use custom route ? 
> 
> `/*A! route(employees_0000.student_0000,employees_0000.student_0007) */ SELECT * FROM student where uid=32`

I think it should follow the custom route, and will be redirected to 0000.

File: pkg/runtime/optimize/hints.go (hunk 3 above)
Comment (ZLBer) [reply pkg/runtime/optimize/hints.go:112]: @binbin0325  what's the meaning of this?

**Review by jjeffcaii — APPROVED** [review]

File: pkg/runtime/optimize/hints.go [hunk 5]
```diff
@@ -62,6 +65,17 @@ func validate(hints []*hint.Hint) error {
 			nodeType = v.Type
 		}
 	}
+	//validate TypeRoute
+	if shardingType == hint.TypeRoute {
+		for _, h := range hints {
```
Comment (binbin0325) [inline pkg/runtime/optimize/hints.go:70]: no need to traverse the hints again

File: pkg/runtime/optimize/hints.go (hunk 3 above)
Comment (binbin0325) [reply pkg/runtime/optimize/hints.go:123]: only routes for custom hints are processed

**Review by binbin0325 — APPROVED** [review]

**Pull request merged**